File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {boolean[]} hasApple
 * @return {number}
 */
var minTime = function(n, edges, hasApple) {
    const graph = buildGraph(edges);
    
    const checked = new Set;
    const mustVisitMemo = [];
    
    const findMustVisitNodes = (node) => {
        if(mustVisitMemo[node]!==undefined) return;
        
        if(checked.has(node)) return;
        checked.add(node);
        
        let res = false;
        
        if(hasApple[node]){
            res = true;
        }
        
            for(let neighbor of graph[node]){
                if(findMustVisitNodes(neighbor)){
                    // mustVisitMemo[node] = true;
                    // return true;
                    res = true;
                }
            }
        
        

        
        
        mustVisitMemo[node] = res;
        return res;
    }

    findMustVisitNodes(0);
    console.log(mustVisitMemo)
    let count = -1;
    for(let node of mustVisitMemo){
        if(node) count++;
    }
    return Math.max(count*2,0);
};

const buildGraph = function(edges){
    // Make an adjacency matrix
    // index is node, element is an array of all neighbors of that node
    const res = [];
    for(let edge of edges){
        if(!res[edge[0]]) res[edge[0]] = [];
        if(!res[edge[1]]) res[edge[1]] = [];
        res[edge[0]].push(edge[1]);
        res[edge[1]].push(edge[0]);
    }
    return res;
}

File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        if not len([bool for bool in hasApple if bool]): return 0
        
        graph = defaultdict(list)
        for [a,b] in edges:
            graph[a].append(b)
            graph[b].append(a)
            
        seen = set()
        must = set()
        
        def dfs(node):
            if node in seen: return False
            seen.add(node)
            
            res = hasApple[node]
            for neighbor in graph[node]:
                if dfs(neighbor): res = True
            
            if res:
                must.add(node)
            return res
        dfs(0)
        
        return (len(must)-1)*2